datafusion/src/physical_optimizer/aggregate_statistics.rs

```diff
@@ -0,0 +1,384 @@
+// Licensed to the Apache Software Foundation (ASF) under one
+// or more contributor license agreements.  See the NOTICE file
+// distributed with this work for additional information
+// regarding copyright ownership.  The ASF licenses this file
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License.  You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+//! Utilizing exact statistics from sources to avoid scanning data
+use std::sync::Arc;
+
+use arrow::datatypes::Schema;
```

Comment:
This file seems to have substantial changes from eef5e2d10e60652aa498925fae3f8c688f5b99fe src/optimizer/aggregate_statistics.rs -- it is not just a rename. I am mostly noting this for other reviewers and my own notes to review it more carefully

Reply:
Agreed. @alamb @houqp @Dandandan @andygrove... this change is really **the gist of this PR**, because this is where we can really judge whether it is a good idea or not to move CBOs to the physical plan. If you compare this optimization with the one originally performed on the logical plan, you see that it becomes a bit more complex because the physical plan can take many more shapes (vtable instead of enum) and is usually larger than the logical one.

Reply:
The code is indeed more complex than the logical optimizer. I feel like moving CBO into physical plane is the right move from a design's point of view. The vtable is hard to avoid because we want to make it extensible by 3rd party plugins.

So perhaps what we can do as a follow up of this PR is to come up with better helper function/macro to make it more readable and maintainable. @Dandandan 's https://github.com/apache/arrow-datafusion/pull/441 and @alamb 's `ExprRewriter` are good examples of such constructs.

Reply:
I went over the logic in this module carefully and I like it and agree with @houqp